model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java

```diff
@@ -856,13 +857,21 @@ public RoleModel getRealmRole(RealmModel realm, String name) {
         if (query == null) {
             Long loaded = cache.getCurrentRevision(cacheKey);
             RoleModel model = getRoleDelegate().getRealmRole(realm, name);
-            if (model == null) return null;
-            query = new RoleListQuery(loaded, cacheKey, realm, model.getId());
+            if (model == null) {
+                // caching empty results will speed up the policy evaluation which tries to look up the role by name and ID
+                query = new RoleListQuery(loaded, cacheKey, realm, Set.of());
+            } else {
+                query = new RoleListQuery(loaded, cacheKey, realm, model.getId());
+            }
             logger.tracev("adding realm role cache miss: client {0} key {1}", realm.getName(), cacheKey);
             cache.addRevisioned(query, startupRevision);
             return model;
         }
-        RoleModel role = getRoleById(realm, query.getRoles().iterator().next());
+        Iterator<String> iterator = query.getRoles().iterator();
```

Comment:
line 904 may need the same logic. 
I would probably add a method `getAny` to `RoleListQuery`.

Reply:
I would be tempted to make it `getOneOrNoRole` and make it throw an exception when there is more than one. And then I am thinking this is overengineered. Now I'm undecided, so I am not sure if I should add it. 
Besides, also added the caching of no entry for client role by name.

Reply:
Creating a new class to cache the query by name would be better. 
The PR looks good to me.

Reply:
Picked up the suggestion for a new class in the latest commit to this PR. Please have a look. The `getRoles()` is still needed to comply with the interface.